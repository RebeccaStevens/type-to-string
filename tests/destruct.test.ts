import dedent from "dedent";
import { describe, it } from "vitest";

import { runTests } from "./helpers";

describe("destructuring arrays", () => {
  it.each([
    dedent`
      const [a, b, ...c] = [1, 2, 3, 4];"
      type Test = typeof a;
    `,
    dedent`
      const [a, b, ...c] = [1, 2, 3, 4];"
      type Test = typeof b;
    `,
    dedent`
      const [a, b, ...c] = [1, 2, 3, 4];
      type Test = typeof c;
    `,
  ])("%s", (code) => {
    runTests(code);
  });
});

describe("destructuring objects", () => {
  it.each([
    dedent`
      const { a, b, ...c } = { a: 1, b: 2, d: 4, e: 5 };"
      type Test = typeof a;
    `,
    dedent`
      const { a, b, ...c } = { a: 1, b: 2, d: 4, e: 5 };"
      type Test = typeof b;
    `,
    dedent`
      const { a, b, ...c } = { a: 1, b: 2, d: 4, e: 5 };
      type Test = typeof c;
    `,
  ])("%s", (code) => {
    runTests(code);
  });
});
