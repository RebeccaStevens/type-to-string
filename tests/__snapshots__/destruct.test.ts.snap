// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`destructuring arrays > const [a, b, ...c] = [1, 2, 3, 4];
type Test = typeof c; 1`] = `
{
  "typeAlias": null,
  "typeReference": null,
  "typeToString": "[number, number]",
}
`;

exports[`destructuring arrays > const [a, b, ...c] = [1, 2, 3, 4];
type Test = typeof c; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "typeof c",
  "typeNodeToString": "[number, number]",
  "typeReferenceNode": null,
}
`;

exports[`destructuring arrays > const [a, b, ...c] = [1, 2, 3, 4];"
type Test = typeof a; 1`] = `
{
  "typeAlias": null,
  "typeReference": null,
  "typeToString": "number",
}
`;

exports[`destructuring arrays > const [a, b, ...c] = [1, 2, 3, 4];"
type Test = typeof a; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "typeof a",
  "typeNodeToString": "number",
  "typeReferenceNode": null,
}
`;

exports[`destructuring arrays > const [a, b, ...c] = [1, 2, 3, 4];"
type Test = typeof b; 1`] = `
{
  "typeAlias": null,
  "typeReference": null,
  "typeToString": "number",
}
`;

exports[`destructuring arrays > const [a, b, ...c] = [1, 2, 3, 4];"
type Test = typeof b; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "typeof b",
  "typeNodeToString": "number",
  "typeReferenceNode": null,
}
`;

exports[`destructuring objects > const { a, b, ...c } = { a: 1, b: 2, d: 4, e: 5 };
type Test = typeof c; 1`] = `
{
  "typeAlias": null,
  "typeReference": null,
  "typeToString": "{ d: number; e: number; }",
}
`;

exports[`destructuring objects > const { a, b, ...c } = { a: 1, b: 2, d: 4, e: 5 };
type Test = typeof c; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "typeof c",
  "typeNodeToString": "{ d: number; e: number; }",
  "typeReferenceNode": null,
}
`;

exports[`destructuring objects > const { a, b, ...c } = { a: 1, b: 2, d: 4, e: 5 };"
type Test = typeof a; 1`] = `
{
  "typeAlias": null,
  "typeReference": null,
  "typeToString": "number",
}
`;

exports[`destructuring objects > const { a, b, ...c } = { a: 1, b: 2, d: 4, e: 5 };"
type Test = typeof a; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "typeof a",
  "typeNodeToString": "number",
  "typeReferenceNode": null,
}
`;

exports[`destructuring objects > const { a, b, ...c } = { a: 1, b: 2, d: 4, e: 5 };"
type Test = typeof b; 1`] = `
{
  "typeAlias": null,
  "typeReference": null,
  "typeToString": "number",
}
`;

exports[`destructuring objects > const { a, b, ...c } = { a: 1, b: 2, d: 4, e: 5 };"
type Test = typeof b; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "typeof b",
  "typeNodeToString": "number",
  "typeReferenceNode": null,
}
`;
